Mathlib/FieldTheory/SeparablyGenerated.lean

```diff
@@ -0,0 +1,407 @@
+      ← Set.image_singleton, ← hai, ← Set.image_union, ← Set.image_univ, Set.compl_union_self]
+  rw [← AlgEquiv.Algebra.isSeparable_iff IntermediateField.topEquiv, ← hk',
+    IntermediateField.isSeparable_adjoin_simple_iff_isSeparable]
+  have hk'' : Algebra.adjoin k (a '' {i}ᶜ) ≤ k'.toSubalgebra :=
+    hai ▸ IntermediateField.algebra_adjoin_le_adjoin _ _
+  let F'' : Polynomial k' := F'.mapAlgHom (B := k') (Subalgebra.inclusion hk'')
+  -- By gauss' lemma, `F` is still irrreducible over `k(x₁,...,xᵢ₋₁, xᵢ₊₁,...,xₙ₊₁)`.
+  have hF''irr : Irreducible F'' := by
+    have inst : NormalizedGCDMonoid (MvPolynomial { b // b ≠ i } k) := Nonempty.some inferInstance
+    have : Irreducible (F.mapAlgHom (IsScalarTower.toAlgHom k _ _)) :=
+      (hFIrr.isPrimitive hF0).irreducible_iff_irreducible_map_fraction_map
+        (K := FractionRing (MvPolynomial { b // b ≠ i } k)).mp hFIrr
+    convert this.map (Polynomial.mapAlgEquiv Hi.1.aevalEquivField) using 1
+    simp only [F', Polynomial.mapAlgHom_comp, ← AlgEquiv.coe_algHom, AlgEquiv.toAlgHom_eq_coe,
+      ← AlgHom.comp_apply, Polynomial.mapAlgEquiv_toAlgHom, F'']
+    congr
+    ext j
+    simp only [ne_eq, Equiv.optionSubtypeNe_apply, Option.casesOn'_some, AlgHom.coe_comp,
+      Function.comp_apply, aeval_X]
+    show a _ = Hi.1.aevalEquivField (algebraMap _ _ _)
+    rw [Hi.1.aevalEquivField_algebraMap_apply_coe, aeval_X]
+  have hF''ai : Polynomial.aeval (a i) F'' = 0 := by
+    simpa only [Polynomial.coe_mapAlgHom, Polynomial.aeval_def, Polynomial.eval₂_map, F''] using hF'
+  have hF''0 : F'' ≠ 0 := (map_ne_zero_iff _ (Polynomial.map_injective _
+    (Subalgebra.inclusion_injective hk''))).mpr fun e ↦ by simp [e] at H
+  -- And by the existence of `Xᵢᵈ` with `p ∤ d`, it is separable.
+  by_contra Hsep
+  have instExpChar : ExpChar k' p := expChar_of_injective_algebraMap (algebraMap k k').injective _
+  have : CharP k' p := by
+    cases instExpChar
+    · cases hp.ne_one rfl
+    · assumption
+  obtain ⟨g, hg, e⟩ := (((minpoly k' (a i)).separable_or p (minpoly.irreducible
+    (isAlgebraic_iff_isIntegral.mp ⟨F'', hF''0, hF''ai⟩))).resolve_left Hsep).2
+  obtain ⟨d, hpd, hF'd⟩ := H
+  replace e := congr(Polynomial.coeff $e d)
+  rw [← minpoly.eq_of_irreducible hF''irr hF''ai, Polynomial.coeff_mul_C,
+    Polynomial.coeff_expand hp.pos, if_neg hpd, eq_mul_inv_iff_mul_eq₀ (by simpa using hF''0),
+    zero_mul, eq_comm] at e
+  exact (Subalgebra.inclusion_injective hk'').ne_iff.mpr hF'd (by simpa [F''] using e)
+
+open IntermediateField
+
+include hp H in
+lemma exists_isTranscendenceBasis_and_isSeparable_of_linearIndepOn_pow_aux
+    (t : Set K) (x : K) (ht : IsTranscendenceBasis k ((↑) : t → K)) (hxt : x ∉ t) :
+    ∃ (i : ↑(insert x t)), (IsTranscendenceBasis k fun (j : {j // j ≠ i}) ↦ j.1.1) ∧
+      IntermediateField.adjoin k (insert x ↑t : Set K) ≤ (separableClosure
+        (IntermediateField.adjoin k (Subtype.val '' {i}ᶜ)) K).restrictScalars _ := by
```

Comment:
Why is this statement not
```suggestion
lemma exists_isTranscendenceBasis_and_isSeparable_of_linearIndepOn_pow_aux
    (t : Set K) (x : K) (ht : IsTranscendenceBasis k ((↑) : t → K)) (hxt : x ∉ t) :
    ∃ i ∈ insert x t, (IsTranscendenceBasis k ((↑) : ↥(insert x t \ {i}) → K)) ∧
      IntermediateField.adjoin k (insert x t) ≤ (separableClosure
        (IntermediateField.adjoin k (insert x t \ {i})) K).restrictScalars _ := by
```
?

Reply:
I suspect this would be easier to prove if we generalize the first main result to
```
variable (a : ι → K) (s : Set ι) (n : ι) (hs : AlgebraicIndepOn k a s)
variable (ha' : ¬ AlgebraicIndepOn k a (insert n s))

lemma exists_algebraicIndepOn_and_isSeparable_of_linearIndepOn_pow_of_adjoin_eq_top :
    ∃ i : ι, AlgebraicIndepOn k a (insert n s \ {i}) ∧
      IsSeparable (IntermediateField.adjoin k (a '' (insert n s \ {i}))) (a i) := by
  sorry
```